### ISSUE TITLE
[Suggestion] Intégrer Mermaid, qui permet de faire des graphes à partir de Markdown

### ISSUE THREAD

**Issue by SpaceFox** [zestedesavoir/zds-site#4715]
Un truc pratique à faire dans un tuto mais toujours un peu compliqué, c'est d'ajouter des graphiques (diagrammes de décision, etc).

C'est là qu'on pourrait ajouter [Mermaid](https://mermaidjs.github.io), qui permet de créer des graphes directement à partir du Markdown (cf [les démos](https://mermaidjs.github.io/demos.html)).

**Avantages**

- C'est prévu pour s'intégrer dans le markdown donc dans le site.
- C'est libre (licence MIT).
- C'est un JS à intégrer : aucune surcharge pour le serveur.
- Les graphes seront versionnés comme tout texte.
- Les designs des graphes seront cohérent à travers tout le site.

**Inconvénients**

- C'est du JS, donc (relativement) lourd pour le client. J'imagine que le transfert des fichiers JS eux-même est compensé par le fait qu'il n'y a plus d'images de graphes à charger. On peut imaginer un _flag_ sur les tutos qui en ont besoin pour éviter de le charger partout.
- Je ne sais pas s'il est facile à intégrer à notre markdown custom.

-----------------

Retours du topic :

- Le JS associé est assez lourd (~800 ko minifié) donc il ne doit être activé (par détection auto) que lorsque nécessaire.
- Il y aurait moyen de faire une v2 qui pré-génèrerait des SVG.

Sujet : https://zestedesavoir.com/forums/sujet/9491/suggestion-integrer-mermaid-qui-permet-de-faire-des-graphes-a-partir-de-markdown/
*Envoyé depuis Zeste de Savoir*

**Comment by artragis** [comment]
C'est plutôt un truc à gérer côté zmarkdown. il y a déjà une issue à ce propos.
https://github.com/zestedesavoir/zmarkdown/issues/171

**Comment by SpaceFox** [comment]
Les deux, non ?

**Comment by artragis** [comment]
Non, zmd génère un svg et basta, faut "juste" coder le plugin qui prend un code mermaid et te sort un svg inliné.

**Comment by SpaceFox** [comment]
Alors attention, on ne parle pas de la même intégration.

Le SVG inliné craché par markdown, c'est la version « avancée ». Si c'est plus simple, on a peut-être tout intérêt à faire la version simple, à savoir : les blocs Mermaid sont conservés à l'identique et entourés d'un `<div class="mermaid">…</div>` et on passe le Mermaid-JS dessus. Cf https://mermaidjs.github.io/usage.html

Le jeu dans cette proposition, c'est vraiment de proposer l'outil aux auteurs le plus simplement possible en terme de dev.

**Comment by artragis** [comment]
> Le SVG inliné craché par markdown, c'est la version « avancée ». 

avec zmd c'est la version "naturelle". et ça évite d'avoir un proco à 100% pour générer les diagrammes à chaque fois, surtout dans les pages dont l'arbre dom est déjà bien touffu.

**Comment by SpaceFox** [comment]
Je ne sais pas : je dis surtout qu'il faut faire attention à ne pas laisser ce genre d'idée en plan parce qu'on essaie de faire une intégration « plus efficace ».

**Comment by artragis** [comment]
Le temps de développement sera le même ou presque (en fait le principal delta viendra du fait qu'avec la génération du svg côté serveur on va surtout permettre d'exporter le tout en latex donc faut l'intégrer à rebber). 

L'utilisation de mermaid dans notre site avec la solution "on place le div class=mermaid" n'est pas si évident car il faut modifier pythonzmd pour creer un exception spécial mermaid, puis réintégrer cette modif ici puis se demander ce qu'on va faire pour les pdf.

Avec l'intégration zmd, on l'intègre dans zmd et basta. C'est le boulot de zmd de gérer le markdown de zds après tout.

**Comment by SpaceFox** [comment]
Ha, c'est parce que j'ai oublié qu'on a un MD spécial qui commente toute
forme de HTML qu'on essaie de lui faire ingérer – ici ça nous complique en
effet sérieusement la vie.

**Comment by artragis** [comment]
PS: ajoutons que logiquement le plan d'attaque c'est

v26 <=> py3 + quelques débug
v27 <=> zmd + quelques debug
v28 <=> reprise des gros chantiers fonctionnels (ping @sandhose @gcodeur)

**Comment by nils-van-zuijlen** [comment]
Certes, l'implémentation back de mermaid doit se faire dans zmd, mais il faudrait avoir un éditeur qui permette d'intégrer facilement des graphes mermaid. Sinon, il ne sera presque jamais utilisé.

**Comment by artragis** [comment]
oui comme je le disais dans ma chronique. Comme la v28 est sensé amener une nouvelle interface de rédaction ça sera sûrement le moment.

**Comment by artragis** [comment]
Pour l'instant c'est en pause à cause d'un problème de CI qui passe pas. @vhf avait proposé un moyen d'améliorer la lib remark-mermaid pour qu'elle soit intégrable dans zmarkdown mais aux dernières nouvelles travis échoue pour des raisons inconnues.
